### ISSUE TITLE
transaction table foreign keys were not successful.

### ISSUE THREAD

**Issue by girl817** [Money-Manager-SaaS/mmss-backend#21]
transaction table foreign keys were not successful. check the code below. @tim-hub 

`> mmss-backend@0.0.1 seed /Users/cd/Desktop/mmss-backend
> ts-node src/db/seeder.ts

/Users/cd/Desktop/mmss-backend/demo.db undefined find db path
(sequelize) Warning: SQLite does not support 'INTEGER' with UNSIGNED or ZEROFILL. Plain 'INTEGER' will be used instead. 
>> Check: https://www.sqlite.org/datatype3.html
Executing (default): CREATE TABLE IF NOT EXISTS `accounts` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `name` VARCHAR(128), `amount` DECIMAL NOT NULL, `currency` VARCHAR(255) NOT NULL DEFAULT 'USD', `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL, `accountID` INTEGER REFERENCES `transactions` (`id`) ON DELETE SET NULL ON UPDATE CASCADE, `toAccountID` INTEGER REFERENCES `transactions` (`id`) ON DELETE SET NULL ON UPDATE CASCADE);
Executing (default): PRAGMA INDEX_LIST(`accounts`)
Executing (default): CREATE TABLE IF NOT EXISTS `transactions` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `transferType` INTEGER DEFAULT -1, `amount` DECIMAL NOT NULL, `accountID` INTEGER NOT NULL, `toAccountID` INTEGER, `note` VARCHAR(256), `date` DATETIME, `categoryID` INTEGER, `payeeID` INTEGER, `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL);
Executing (default): PRAGMA INDEX_LIST(`transactions`)
Executing (default): CREATE TABLE IF NOT EXISTS `categories` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `name` VARCHAR(128), `description` VARCHAR(256), `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL, `categoryID` INTEGER REFERENCES `transactions` (`id`) ON DELETE SET NULL ON UPDATE CASCADE);
Executing (default): PRAGMA INDEX_LIST(`categories`)
Executing (default): CREATE TABLE IF NOT EXISTS `users` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `username` VARCHAR(256) NOT NULL, `email` VARCHAR(256), `last_login` DATETIME, `date_login` DATETIME, `passwordhash` VARCHAR(256) NOT NULL, `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL, `userid` INTEGER REFERENCES `ledgers` (`id`) ON DELETE SET NULL ON UPDATE CASCADE);
Executing (default): PRAGMA INDEX_LIST(`users`)
Executing (default): CREATE TABLE IF NOT EXISTS `ledgers` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `userid` INTEGER, `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL);
Executing (default): PRAGMA INDEX_LIST(`ledgers`)
Executing (default): INSERT INTO `accounts` (`id`,`name`,`amount`,`currency`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5);
Executing (default): INSERT INTO `accounts` (`id`,`name`,`amount`,`currency`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5);
Executing (default): INSERT INTO `categories` (`id`,`name`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3);
Executing (default): INSERT INTO `categories` (`id`,`name`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3);
Executing (default): INSERT INTO `transactions` (`id`,`transferType`,`amount`,`accountID`,`note`,`categoryID`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5,$6,$7);
Executing (default): INSERT INTO `transactions` (`id`,`transferType`,`amount`,`accountID`,`toAccountID`,`note`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5,$6,$7);
Executing (default): INSERT INTO `transactions` (`id`,`transferType`,`amount`,`accountID`,`note`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5,$6);
Executing (default): INSERT INTO `users` (`id`,`username`,`email`,`last_login`,`date_login`,`passwordhash`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5,$6,$7);
Executing (default): INSERT INTO `users` (`id`,`username`,`email`,`last_login`,`date_login`,`passwordhash`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4,$5,$6,$7);
Executing (default): INSERT INTO `ledgers` (`id`,`userid`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3);
db seeding done
`

**Comment by tim-hub** [comment]
can you past more detail?

I did see any thing wrong in the log.

do you mean the db demo.db is not updated?

**Comment by cdhit** [comment]
No foreign key defined in transaction table create script. 

Regards,
Don

**Comment by cdhit** [comment]
No foreign key defined from below. 

Executing (default): CREATE TABLE IF NOT EXISTS transactions (idINTEGER PRIMARY KEY AUTOINCREMENT, transferTypeINTEGER DEFAULT -1, amountDECIMAL NOT NULL, accountIDINTEGER NOT NULL, toAccountIDINTEGER, note VARCHAR(256), dateDATETIME, categoryID INTEGER, payeeID INTEGER, createdAtDATETIME NOT NULL, updatedAtDATETIME NOT NULL);
Executing (default): PRAGMA INDEX_LIST(transactions)



Regards,
Don

**Comment by tim-hub** [comment]
will take some time to debug this, cannot help now

suggestion, 

1. revert master back to normal
2. try to follow https://github.com/Money-Manager-SaaS/mmss-backend/pull/22/files the review to do some changes, upper, lowercase might cause this problem

**Comment by tim-hub** [comment]
> No foreign key defined from below. Executing (default): CREATE TABLE IF NOT EXISTS transactions (idINTEGER PRIMARY KEY AUTOINCREMENT, transferTypeINTEGER DEFAULT -1, amountDECIMAL NOT NULL, accountIDINTEGER NOT NULL, toAccountIDINTEGER, note VARCHAR(256), dateDATETIME, categoryID INTEGER, payeeID INTEGER, createdAtDATETIME NOT NULL, updatedAtDATETIME NOT NULL); Executing (default): PRAGMA INDEX_LIST(transactions) Regards, Don
> […](#)
> On 5/05/2020, at 8:50 PM, Tim ***@***.***> wrote: can you past more detail? I did see any thing wrong in the log. do you mean the db demo.db is not updated? — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or unsubscribe.

which table/model?

**Comment by tim-hub** [comment]
BTW, do not rush to merge

**Comment by tim-hub** [comment]
problem is you did not add relation to the items you created.

comparing these two
```

  const withdraw = await Transaction.create({
    transferType: -1,
    amount: 10,
    accountID: cash.id,
    categoryID: food.id,
    note: 'a spending on food'
  });


```

```
  const tom = await User.create({
    username: 'tom',
    email: 'tom.c@Kmail.com',
    last_login: '2011/10/19',
    date_login: '2018/03/18',
    passwordhash: '90d0dke888dje3j'
  });


  const supermarket = await Ledger.create({
    userid:7870009,
    ledgername: 'supermarket',
  });
```

you never used your own user, I mentioned this in the PR review, not hard code userID

**Comment by tim-hub** [comment]
too buggy, will revert

**Comment by tim-hub** [comment]
- please read the review comments and fix the issues I mentioned there first,
- not only code style, but also the logic 
- https://github.com/Money-Manager-SaaS/mmss-backend/issues/21#issuecomment-623942807 I hope this comment solved your question


- please never rush to merge.

https://github.com/Money-Manager-SaaS/mmss-backend/pull/22

**Comment by cdhit** [comment]
fixed in https://github.com/Money-Manager-SaaS/mmss-backend/pull/25

**Comment by cdhit** [comment]
fixed in #25. Solved.

**Comment by cdhit** [comment]
fixed in #25. Solved.
